contracts/OptimisticArbitrator.sol

```diff
@@ -0,0 +1,199 @@
+            msg.sender,
+            address(this),
+            request.bond + request.finalFee + 2 * finalFeeRise
+        );
+
+        request.currency.approve(address(oo), 2 * (request.bond + request.finalFee + finalFeeRise) + request.reward);
+
+        bytes memory disputeAncillaryData = _getDisputeAncillaryData(requestId);
+        oo.requestPrice(priceIdentifier, timestamp, disputeAncillaryData, request.currency, request.reward);
+        oo.setBond(priceIdentifier, timestamp, disputeAncillaryData, request.bond);
+        oo.proposePriceFor(request.proposer, address(this), priceIdentifier, timestamp, disputeAncillaryData, 1e18);
+        oo.disputePriceFor(msg.sender, address(this), priceIdentifier, timestamp, disputeAncillaryData);
+
+        // If the final fee has decreased, refund the excess to the disputer & proposer.
+        uint256 balanceToRefund = request.currency.balanceOf(address(this)) > initialBalance
+            ? request.currency.balanceOf(address(this)) - initialBalance
+            : 0;
+        if (balanceToRefund > 0) request.currency.safeTransfer(msg.sender, balanceToRefund);
+    }
+
+    function settleAndGetPrice(uint256 timestamp, bytes memory ancillaryData) public returns (uint256[4] memory) {
+        bytes32 requestId = _getId(msg.sender, timestamp, ancillaryData);
+        Request storage request = requests[requestId];
+        require(address(request.currency) != address(0), "Price not requested");
+        require(request.proposer != address(0), "No proposed price to settle");
+
+        if (request.disputer != address(0)) {
+            require(
+                oo.settleAndGetPrice(priceIdentifier, timestamp, _getDisputeAncillaryData(requestId)) == 1e18,
+                "Price not resolved correctly"
+            );
+        } else {
+            require(uint64(getCurrentTime()) > request.expirationTime, "Proposal not passed liveness");
+            request.currency.safeTransfer(request.proposer, request.bond + request.finalFee + request.reward);
+        }
+
+        request.settled = true;
+
+        return request.proposedPrice;
+    }
+
+    function getPrice(uint256 timestamp, bytes memory ancillaryData) public view returns (uint256[4] memory) {
+        Request storage request = requests[_getId(msg.sender, timestamp, ancillaryData)];
+        require(request.settled == true, "Request not settled");
+        return request.proposedPrice;
+    }
+
+    // This function must return a bytes value with length that is shorter than or equal to oo.OO_ANCILLARY_DATA_LIMIT=8139
+    // See the OptimisticOracleV2 implementation for more details.
+    function _getDisputeAncillaryData(bytes32 queryId) public view returns (bytes memory) {
+        return
+            abi.encodePacked(
+                'q: "Is the proposed price to the request with ID: ',
+                queryId,
+                " in the following contract: ",
+                address(this),
+                ' valid?"'
+            );
+    }
```

Comment:
This is not typical but will help a lot with the explanation in the tutorial. would be cool to show when you are going through how to use the contract via remix to get the user to decode the ancillary data via a website like https://web3-type-converter.onbrn.com/ or equivalent to show the structure here in.